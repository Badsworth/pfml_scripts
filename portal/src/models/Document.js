/* eslint sort-keys: ["error", "asc"] */
/**
 * @file Document (AKA File) model and enum values
 */
import BaseModel from "./BaseModel";
import LeaveReason from "./LeaveReason";

class Document extends BaseModel {
  get defaults() {
    return {
      application_id: null,
      content_type: null,
      created_at: null,
      description: null,
      document_type: null,
      fineos_document_id: null,
      name: null,
      user_id: null,
    };
  }
}

/**
 * Enums for Document `document_type` field.  In the `certification` object, `certificationForm` is only used for uploads of certification forms; the API then sets the plan
 * proof based on leave reason.  The other `certification` types are used when we retrieve documents from FINEOS.
 * @enum {string}
 */
export const DocumentType = {
  approvalNotice: "Approval Notice",
  certification: {
    certificationForm: "Certification Form",
    [LeaveReason.activeDutyFamily]: "Military exigency form",
    [LeaveReason.care]: "Care for a family member form",
    [LeaveReason.bonding]: "Child bonding evidence form",
    [LeaveReason.medical]: "Own serious health condition form",
    [LeaveReason.pregnancy]: "Pregnancy/Maternity form",
    [LeaveReason.serviceMemberFamily]: "Care for a family member form",
    medicalCertification: "State managed Paid Leave Confirmation", // TODO (CP-2029): Remove this legacy type once claims filed before 7/1/2021 are adjudicated
  },
  denialNotice: "Denial Notice",
  identityVerification: "Identification Proof",
  medicalCertification: "State managed Paid Leave Confirmation",
  requestForInfoNotice: "Request for more Information",
  withdrawalNotice: "Withdrawal Notice"
};

export default Document;
