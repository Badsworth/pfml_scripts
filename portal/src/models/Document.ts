/* eslint sort-keys: ["error", "asc"] */
/**
 * @file Document (AKA File) model and enum values
 */
import LeaveReason from "./LeaveReason";

const CertificationType = {
  certificationForm: "Certification Form",
  [LeaveReason.activeDutyFamily]: "Military exigency form",
  [LeaveReason.care]: "Care for a family member form",
  [LeaveReason.bonding]: "Child bonding evidence form",
  [LeaveReason.medical]: "Own serious health condition form",
  [LeaveReason.pregnancy]: "Pregnancy/Maternity form",
  [LeaveReason.serviceMemberFamily]: "Care for a family member form",
  medicalCertification: "State managed Paid Leave Confirmation", // TODO (CP-2029): Remove this legacy type once claims filed before 7/1/2021 are adjudicated
} as const;

const OtherDocumentType = {
  appealAcknowledgment: "Appeal Acknowledgment",
  approvalNotice: "Approval Notice",
  denialNotice: "Denial Notice",
  identityVerification: "Identification Proof",
  medicalCertification: "State managed Paid Leave Confirmation",
  requestForInfoNotice: "Request for more Information",
  withdrawalNotice: "Pending Application Withdrawn",
} as const;

export type DocumentTypeEnum =
  | typeof CertificationType[keyof typeof CertificationType]
  | typeof OtherDocumentType[keyof typeof OtherDocumentType];

/**
 * Enums for Document `document_type` field.  In the `certification` object, `certificationForm` is only used for uploads of certification forms; the API then sets the plan
 * proof based on leave reason.  The other `certification` types are used when we retrieve documents from FINEOS.
 * @enum {string}
 */
export const DocumentType = {
  certification: { ...CertificationType },
  ...OtherDocumentType,
} as const;
