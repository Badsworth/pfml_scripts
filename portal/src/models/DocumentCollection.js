import BaseCollection from "./BaseCollection";

export default class DocumentCollection extends BaseCollection {
  get idProperty() {
    return "fineos_document_id";
  }

  /**
   * Get only documents associated with a given Application
   * @param {string} application_id - ID of the target Application
   * @returns {Document[]}
   */
  filterByApplication(application_id) {
    return this.items.filter((item) => item.application_id === application_id);
  }
}
