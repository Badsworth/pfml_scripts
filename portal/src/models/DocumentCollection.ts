import BaseCollection from "./BaseCollection";
import BenefitsApplicationDocument from "./BenefitsApplicationDocument";
import ClaimDocument from "./ClaimDocument";

export default class DocumentCollection extends BaseCollection<
  BenefitsApplicationDocument | ClaimDocument
> {
  get idProperty() {
    return "fineos_document_id";
  }

  /**
   * Get only documents associated with a given Application
   */
  filterByApplication(application_id: string) {
    return this.items.filter((item) => {
      if (item instanceof BenefitsApplicationDocument) {
        return item.application_id === application_id;
      }
      return false;
    });
  }
}
