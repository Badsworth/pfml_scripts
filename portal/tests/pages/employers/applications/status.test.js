import Document, { DocumentType } from "../../../../src/models/Document";
import {
  MockEmployerClaimBuilder,
  renderWithAppLogic,
} from "../../../test-utils";
import DocumentCollection from "../../../../src/models/DocumentCollection";
import Status from "../../../../src/pages/employers/applications/status";

jest.mock("../../../../src/hooks/useAppLogic");

const CLAIM = new MockEmployerClaimBuilder()
  .status("Undecided")
  .completed()
  .create();
const DOCUMENTS = new DocumentCollection([
  new Document({
    content_type: "application/pdf",
    created_at: "2020-01-02",
    document_type: DocumentType.approvalNotice,
    fineos_document_id: "fineos-id-1",
    name: "Approval notice doc",
  }),
  new Document({
    content_type: "application/pdf",
    created_at: "2020-02-03",
    document_type: DocumentType.denialNotice,
    fineos_document_id: "fineos-id-2",
    name: "Denial notice doc",
  }),
  new Document({
    content_type: "image/png",
    created_at: "2020-04-05",
    document_type: DocumentType.medicalCertification,
    fineos_document_id: "fineos-id-4",
    name: "Medical cert doc",
  }),
  new Document({
    content_type: "image/png",
    created_at: "2020-03-04",
    document_type: DocumentType.requestForInfoNotice,
    fineos_document_id: "fineos-id-3",
    name: "Request for info doc",
  }),
]);

describe("Status", () => {
  const query = { absence_id: "NTN-111-ABS-01" };
  let appLogic;
  let wrapper;

  beforeEach(() => {
    process.env.featureFlags = {
      employerShowDashboard: false,
    };

    ({ appLogic, wrapper } = renderWithAppLogic(Status, {
      employerClaimAttrs: CLAIM,
      props: {
        query,
      },
    }));
  });

  it("renders the page and correct content in Trans component", () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("Trans").dive()).toMatchSnapshot();
  });

  it("shows the claimant name as the title", () => {
    expect(wrapper.find("Title").childAt(0).text()).toEqual(
      "Application status for Jane Doe"
    );
  });

  it("shows lead text for pending claims", () => {
    expect(wrapper.find("Lead").exists()).toEqual(true);
    expect(wrapper.find("Trans").dive()).toMatchSnapshot();
  });

  it("shows lead text for resolved claims", () => {
    const resolvedClaim = new MockEmployerClaimBuilder()
      .status("Approved")
      .completed()
      .create();
    ({ wrapper } = renderWithAppLogic(Status, {
      employerClaimAttrs: resolvedClaim,
      props: {
        query,
      },
    }));

    expect(wrapper.find("Lead").exists()).toEqual(true);
    expect(wrapper.find("Trans").dive()).toMatchSnapshot();
  });

  it("shows the application ID", () => {
    const applicationIdRow = wrapper.find("StatusRow").first();
    expect(applicationIdRow.prop("label")).toEqual("Application ID");
    expect(applicationIdRow.childAt(0).text()).toEqual("NTN-111-ABS-01");
  });

  it("shows the leave type", () => {
    const leaveTypeRow = wrapper.find("StatusRow").at(1);
    expect(leaveTypeRow.prop("label")).toEqual("Leave type");
    expect(leaveTypeRow.childAt(0).text()).toEqual("Medical leave");
  });

  it("shows the leave duration", () => {
    const leaveDurationRow = wrapper.find("StatusRow").last();
    expect(leaveDurationRow.prop("label")).toEqual("Leave duration");
    expect(leaveDurationRow.childAt(0).text()).toEqual("1/1/2021 â€“ 7/1/2021");
  });

  describe("when documents haven't been loaded yet", () => {
    it("only shows the 'Leave details' section title", () => {
      const sectionTitle = wrapper.find("Heading");
      expect(sectionTitle.first()).toStrictEqual(sectionTitle.last());
      expect(sectionTitle.childAt(0).text()).toEqual("Leave details");
    });

    it("loads the documents", () => {
      ({ appLogic } = renderWithAppLogic(Status, {
        employerClaimAttrs: CLAIM,
        props: {
          query,
        },
        render: "mount",
      }));

      expect(appLogic.employers.loadDocuments).toHaveBeenCalledWith(
        query.absence_id
      );
    });
  });

  describe("when there are no documents", () => {
    it("only shows the 'Leave details' section title", () => {
      ({ appLogic } = renderWithAppLogic(Status, {
        employerClaimAttrs: CLAIM,
        props: {
          query,
        },
        render: "mount",
      }));
      appLogic.employers.documents = new DocumentCollection([]);

      const sectionTitle = wrapper.find("Heading");
      expect(sectionTitle.first()).toStrictEqual(sectionTitle.last());
      expect(sectionTitle.childAt(0).text()).toEqual("Leave details");
    });
  });

  describe("when 'employerShowDashboard' is enabled", () => {
    beforeEach(() => {
      process.env.featureFlags = {
        employerShowDashboard: true,
      };

      ({ wrapper } = renderWithAppLogic(Status, {
        employerClaimAttrs: CLAIM,
        props: {
          query,
        },
      }));
    });

    it("displays adjudication status if status is provided", () => {
      const tagComponent = wrapper.find("StatusRow").at(1).dive().find("Tag");
      expect(tagComponent.exists()).toBe(true);
      expect(tagComponent.prop("state")).toEqual("warning");
      expect(tagComponent.prop("label")).toEqual("pending");
    });

    it("hides adjudication status if status is undefined", () => {
      const claimWithoutStatus = new MockEmployerClaimBuilder()
        .completed()
        .create();
      ({ wrapper } = renderWithAppLogic(Status, {
        employerClaimAttrs: claimWithoutStatus,
        props: {
          query,
        },
      }));

      const tagComponent = wrapper.find("StatusRow").at(1).dive().find("Tag");
      expect(tagComponent.exists()).toBe(false);
    });
  });

  describe("when there are documents", () => {
    beforeEach(() => {
      appLogic.employers.documents = DOCUMENTS;
      appLogic.employers.downloadDocument = jest.fn();

      ({ wrapper } = renderWithAppLogic(Status, {
        employerClaimAttrs: CLAIM,
        props: {
          appLogic,
          query,
        },
      }));
    });

    it("shows the appropriate headings", () => {
      const sectionTitles = wrapper
        .find("Heading")
        .map((section) => section.childAt(0).text());
      expect(sectionTitles).toEqual(["Leave details", "Notices"]);
    });

    it("shows only legal documents", () => {
      const documentDivs = wrapper.find("DocumentListItem");
      const entries = documentDivs.map((div) => {
        const title = div.dive().find("p").at(0).text();
        const date = div.dive().find("p").at(1).text();
        return [title, date];
      });

      expect(entries).toEqual([
        ["Approval notice (PDF)", "Posted 1/2/2020"],
        ["Denial notice (PDF)", "Posted 2/3/2020"],
        ["Request for more information", "Posted 3/4/2020"],
      ]);
    });

    it("makes a call to download documents on click", async () => {
      const downloadDocumentSpy = jest
        .spyOn(appLogic.employers, "downloadDocument")
        .mockImplementation(() => Promise.resolve(new Blob()));

      await wrapper
        .find("DocumentListItem")
        .at(0)
        .dive()
        .find("a")
        .simulate("click", {
          preventDefault: () => jest.fn(),
        });

      expect(downloadDocumentSpy).toHaveBeenCalledWith(
        "NTN-111-ABS-01",
        expect.objectContaining({
          content_type: "application/pdf",
          created_at: "2020-01-02",
          document_type: "Approval Notice",
          fineos_document_id: "fineos-id-1",
          name: "Approval notice doc",
        })
      );
    });
  });
});
