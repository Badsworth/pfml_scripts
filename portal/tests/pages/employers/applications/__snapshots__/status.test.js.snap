// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Status renders the page and correct content in Trans component 1`] = `
<Fragment>
  <BackButton />
  <Title>
    Application status for Jane Doe
  </Title>
  <Lead>
    <Trans
      components={
        Object {
          "dfml-regulations-link": <a
            href="https://www.mass.gov/regulations/458-CMR-200-department-of-family-and-medical-leave-1"
            rel="noopener"
            target="_blank"
          />,
        }
      }
      i18nKey="pages.employersClaimsStatus.lead"
      tOptions={
        Object {
          "context": "pending",
        }
      }
    />
  </Lead>
  <Heading
    level="2"
  >
    Leave details
  </Heading>
  <StatusRow
    label="Application ID"
  >
    NTN-111-ABS-01
  </StatusRow>
  <StatusRow
    label="Leave type"
  >
    Medical leave
  </StatusRow>
  <StatusRow
    label="Leave duration"
  >
    1/1/2021 – 7/1/2021
  </StatusRow>
  <StatusRow
    label="Continuous leave"
  >
    1/1/2021 – 6/1/2021
  </StatusRow>
  <StatusRow
    label="Reduced leave schedule"
  >
    2/1/2021 – 7/1/2021
  </StatusRow>
</Fragment>
`;

exports[`Status renders the page and correct content in Trans component 2`] = `
Array [
  "This application is being reviewed by the Department. It was either reviewed by an administrator on your team or the review deadline has passed. No action is required of you.",
  <br
    key="br-1"
  />,
  <br
    key="br-2"
  />,
  "After we make a decision, you'll receive an email with a direct link for more details.",
]
`;

exports[`Status shows lead text for pending claims 1`] = `
Array [
  "This application is being reviewed by the Department. It was either reviewed by an administrator on your team or the review deadline has passed. No action is required of you.",
  <br
    key="br-1"
  />,
  <br
    key="br-2"
  />,
  "After we make a decision, you'll receive an email with a direct link for more details.",
]
`;

exports[`Status shows lead text for resolved claims 1`] = `
Array [
  "A decision has been made for this application. No action is required of you, but you can download a copy of the decision notice for details. Your employee has the right to appeal this decision under Massachusetts regulations (",
  <a
    href="https://www.mass.gov/regulations/458-CMR-200-department-of-family-and-medical-leave-1"
    key="1"
    rel="noopener"
    target="_blank"
  >
    458 CMR 2.14
  </a>,
  ").",
]
`;
